subsys/bluetooth/host/att.c

```diff
@@ -3012,6 +3012,18 @@ static void bt_att_connected(struct bt_l2cap_chan *chan)
 		le_chan->rx.mtu = BT_ATT_DEFAULT_LE_MTU;
 	}
 
+	if (le_chan->psm == BT_EATT_PSM) {
+		if (le_chan->tx.mtu > CONFIG_BT_L2CAP_TX_MTU) {
+			BT_DBG("chan %p (0x%04x): saturating TX MTU to %d",
+			       le_chan, le_chan->tx.cid, CONFIG_BT_L2CAP_TX_MTU);
+		}
+
+		le_chan->tx.mps = MIN(le_chan->tx.mps,
+				      BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU));
+		le_chan->tx.mtu = MIN(le_chan->tx.mtu,
+				      CONFIG_BT_L2CAP_TX_MTU);
```

Comment:
If the remote side requests an EATT with the MPS > BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU) or an MTU > CONFIG_BT_L2CAP_TX_MTU, we should reject it instead. 

If the application/whatever API that can request these values are setting them above, we should reject it. 

Silently truncating the values in the `connected` callback seem to be the wrong place and way of handling this IMO

Reply:
> BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 3, Part G 5.3.1 ATT_MTU
The ATT_MTU for the Enhanced ATT bearer shall be set to the minimum of the
MTU field values of the two devices; these values come from the L2CAP_CREDIT_BASED_CONNECTION_REQ and L2CAP_CREDIT_BASED_CONNECTION_RSP signaling packets or the latest L2CAP_CREDIT_BASED_RECONFIGURE_REQ packets.
Note: The minimum ATT_MTU for an Enhanced ATT bearer is 64 octets.

Both sides shall not send larger packets than the smallest MTU of the two peers. If we reject requests from peers that support a longer MTU, we will end up rejecting a lot of higher-memory devices like phones.

Reply:
Our RX MTU should already have done these buffer size calculations. Can this be simplified to `tx_mtu = min(tx_mtu, rx_mtu)`? That would more directly match the statement from the spec as well.

Reply:
My point was that doing this in the `connected` callback seem to be a late stage to limit the MPS and MTU. Isn't it possible to restrict this at the channel request stage?

Reply:
I think this is the earliest we can. The accept callback `bt_eatt_accept` doesn't yet have the MTU information: it is that function that allocates the channel structure, so all the fields are at 0.
II initially tried to do it there (rather in `att_chan_new` that is called by it).

Reply:
> I think this is the earliest we can. 

I don't think that can be true, since here the `le_chan->tx.mps` and `le_chan->tx.mtu` must already be initialized to some value (if not, then these truncations aren't doing anything). 

So whenever these values are initially initialized to the values, is where these truncations should be. 

In other words, this cannot be the first time the values are set, and the truncations should be done at the first time the values are set.

Reply:
yeah that's done in l2cap.c, and I don't want to have l2cap knowing about EATT-specific stuff (ie the PSM).

Reply:
> yeah that's done in l2cap.c, and I don't want to have l2cap knowing about EATT-specific stuff (ie the PSM).

But the change here
```c
		le_chan->tx.mps = MIN(le_chan->tx.mps,
				      BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU));
		le_chan->tx.mtu = MIN(le_chan->tx.mtu,
				      CONFIG_BT_L2CAP_TX_MTU);
```

Seems to be L2CAP only, right? The PSM is also L2CAP, and is used for any dynamic channels, and not just for EATT, right? 

I don't see why this check isn't better placed in L2CAP.

Reply:
no the PSM is only used for EATT

Reply:
anyway, I'm not going to proceed with the change as-is, as @alwa-nordic identified an issue with this approach (when reconfiguring the channel).
I'm probably going to post a quick-fix PR to the place where the failed assert is, and post another PR introducing the concept of ATT MTU.

Reply:
> no the PSM is only used for EATT

That doesn't sound right. From our L2CAP docs:
```
/** @brief Register L2CAP server.
 *
 *  Register L2CAP server for a PSM, each new connection is authorized using
 *  the accept() callback which in case of success shall allocate the channel
 *  structure to be used by the new connection.
 *
 *  For fixed, SIG-assigned PSMs (in the range 0x0001-0x007f) the PSM should
 *  be assigned to server->psm before calling this API. For dynamic PSMs
 *  (in the range 0x0080-0x00ff) server->psm may be pre-set to a given value
 *  (this is however not recommended) or be left as 0, in which case upon
 *  return a newly allocated value will have been assigned to it. For
 *  dynamically allocated values the expectation is that it's exposed through
 *  a GATT service, and that's how L2CAP clients discover how to connect to
 *  the server.
 *
 *  @param server Server structure.
 *
 *  @return 0 in case of success or negative value in case of error.
 */
int bt_l2cap_server_register(struct bt_l2cap_server *server);
```

We also in e.g. OTS do 
```
static struct bt_l2cap_server l2cap_server = {
	.psm = BT_GATT_OTS_L2CAP_PSM,
	.accept	= l2cap_accept,
};
```

The PSM is used for any dynamic L2CAP channel IIRC

Reply:
I meant the PSM value/id, not the psm as a concept. I've updated the PR to not need it anyway.